### DISCUSSION TITLE
What is "trainable" about the lemmatizer?

### DISCUSSION THREAD

**Discussion by BramVanroy** [explosion/spaCy#8686] [Help: Other Questions] [answered]
I can [see](https://spacy.io/api/lemmatizer) that either a rule-based or lookup lemmatizer can be implemented and that many models use a trained, rule-based lemmatizer. I cannot find what kind of rule-deduction is being done, however. I've been looking for the lemmatizer model [here](https://github.com/explosion/spaCy/tree/master/spacy/ml) but could not find it.

Tl;dr: what kind of lemmatization strategy is used when using `rule`? Tree edit operations, string edit operations, etc.

**Comment by adrianeboyd** [comment, marked as answer]
Hmm, are you referring to having `TrainablePipe` as the base class in the docs? The "Trainable" flag next to that is correct: there is not currently a trainable lemmatizer.

In terms of the docs, the provided base class is a mistake as compared to the source (the base class is currently `Pipe`), but the `Lemmatizer` class has been designed so that it can be extended in the future without changing the API, so it has an unused `Model` parameter.

The current rule-based lemmatizers are all hand-written.

**Reply by BramVanroy** [reply]
It seems that I completely missed the `Trainable: X` part. I just saw the base class and figured it was trainable indeed. 

It would be interesting indeed to have a trainable lemmatizer indeed.

A related question: how does the config know which rules to use when I train a new model? I see in en_core_web_sm the following as the only information about the lemmatizer:

```
[components.lemmatizer]

factory = "lemmatizer"
mode = "rule"
model = null
overwrite = false
```

The same question goes for the `attribute_ruler` which is also not trainable. My guess is that it defaults to a language-specific setting due to `lang = "en"` and then uses built-in spaCy defaults. The question then becomes: what should I do if I want to train a model which has _no_ language defaults in spaCy yet?

**Reply by adrianeboyd** [reply]
There are default rule tables for each lemmatizer mode and languages can override those with custom lemmatizer (French is an example). The tables are loaded from `spacy-lookups-data`, but the size of that package is becoming an issue, so we probably need to additionally support loading from URLs soon.

The `attribute_ruler` settings are pipeline-specific rather than language-specific, since they depend on things like which components are in the pipeline and which tagsets are in the tagger corpus. Right now the source is stored in an internal repo, but you can inspect any rule sets in the pretrained pipelines (`nlp.get_pipe("attribute_ruler").patterns`) and it's identical to the internal files.

If there are aren't any language defaults, then you probably want to define a custom language. You may or may not be able to add a lemmatizer depending on the available resources.

The `attribute_ruler` settings are extremely pipeline-specific. We use some pretty generic rules in all pretrained pipelines to make sure whitespace is tagged consistently and to copy pos to tag for pipelines without taggers (just for convenience for users), but there aren't any general-purpose recommendations here -- it really depends on the pipeline.

**Reply by BramVanroy** [reply]
So from a training perspective I understand that if I use the exact same pipeline as e.g. en_core_web_sm (without ner), I can just use the default lookup tables for the lemmatizer, which will then look in `spacy-lookups-data` for `rule` and `en` rules.

For the `attribute_ruler` I am not sure I understand yet. In the train config of `en_core_web_sm` but also others like `en_core_web_trf` I only see

```
[components.attribute_ruler]
factory = "attribute_ruler"
validate = false
```

So how does these specific pipelines differ/know which rules they need to find? Or, put differently, if I train a model from scratch with the above component, how will the script know to use "the right" rules?

In my case I am working with UD, so I am not sure whether the attribute ruler works on that data as it seems to use the POS tags, which may differ between UD and spaCy models. I did find [the example to train a UD pipeline](https://github.com/explosion/projects/blob/v3/pipelines/tagger_parser_ud/configs/default.cfg), but it does not use an `attribute_ruler` at all. But reading [the documentation](https://spacy.io/usage/linguistic-features/#mappings-exceptions), it seems that an attribute ruler does contain useful things like token exceptions and tags for spaces. So is it more that: UD training does not need an attribute ruler, or more that the attribute_ruler is not robust enough (yet?) to be used with different data?

Thanks

**Reply by adrianeboyd** [reply]
The lemmatizer lookups and attribute ruler patterns would typically be added in the `[initialize]` block. 

https://spacy.io/api/attributeruler#initialize

https://spacy.io/api/lemmatizer#initialize

If lemmatizer `lookups` aren't provided in the initialize block, then it loads the tables from `spacy-lookups-data`.

(In the pretrained pipelines, we train most of the components separately from separate configs and then merge them in a "collate" script, so those configs aren't good examples.)

You never **need** an attribute ruler and there are no default rules, you always have to add the rules that makes sense for your pipeline. The rules all depend on your components and training data, so it's not really possible to provide default rules. But once rules are added, they're serialized with the model, so you only need to do the loading once (which is why it should be in `initialize`).

It doesn't have to do with being robust or not, there just aren't any truly general-purpose rules or exceptions. In your pipeline you might not even want the `SPACE/_SP` tagging that we do by default in the pretrained pipelines.

**Reply by BramVanroy** [reply]
So it is basically a way to catch oddities/artificats/unexpected behaviour of the pipeline? In my case it might be better to just train the pipeline without it and see whether some specific tags have low F1 and why, and see whether some rules can help with that.

So I trained a first UD model with XLM-R on EWT UD and the performance is decent, but indeed lemmata are not predicted very well (F1=0.75). This was just a quick run without understanding much of the config so there are many flaws I am sure, but the tag predictions are quite good.

```
[components.lemmatizer]
factory = "lemmatizer"
mode = "rule"
model = null
overwrite = false
```

This is the only thing I have related to the lemmatizer (as I saw that in the en_core_web_sm config) but there is no initialize block.  Am I correct in that even without an initialize block the lemmatizer is still initialized? And in this case it looks for the "rule" and "en" tables in `spacy-lookups-data`?

If you have other suggestions to improve the lemmatizer I am all ears. It seems that in the [example script](https://github.com/explosion/projects/blob/v3/pipelines/tagger_parser_ud/configs/default.cfg) a lemmatizer was not even included. I am not entirely sure why. Perhaps you noticed the same gap in performance with UD data?

**Reply by adrianeboyd** [reply]
All components are initialized even if there's no block in `[initialize]`.

The rule-based lemmatizer just isn't particularly interesting for that example, which is focused on data conversion and training. The lemmatizer evaluation can depend on a lot of "minor" factors like how to lemmatize proper nouns or pronouns or punctuation, so to improve the performance on a UD corpus, you'd need to customize the rules for that particular corpus. Lemmas are typically not consistent across corpora in the same language, either.

**Reply by BramVanroy** [reply]
That's what I thought. I remember that pronouns used to be lemmatised as -PRON-, although that does not seem to be rule based following the rules that I found in spacy-lookups-data. 

So concretely, if I want to update the lemma rules and lookups according to my own dataset, how can I tell the config to use my custom files? From what I read in [the documentation](https://spacy.io/api/data-formats#config) it seems that this _has_ to be part of the lookups data package:

> Additional lexeme and vocab data from spacy-lookups-data

Can I use a path? My guess is that I can build my custom Lookups and add the required tables with .add_table and then save the whole Lookup `.to_disk`. That file path could then be used in in `initialize.lookups`. Is that correct? The one doubt that I have is that I am not sure whether there are naming conventions for the tables. I.e., how does the trainer (lemmatizer) know which tables it must use in my custom Lookup?

I am sorry about all these questions but I really have tried looking for all this information. The training-config system is very customizable and that is a great thing, but precisely because of its abilities I find it hard to wrap my head around it when I want to do something with custom data/annotations/architectures. I hope you understand.

**Reply by adrianeboyd** [reply]
v3 gets rid of `-PRON-` as a lemma.

The lookups function in `initialize` should return a `Lookups` object, and internally you can load those tables however you want. Simple example:

https://github.com/explosion/spaCy/blob/8233359225df4b959a9d856ef208a66756f73502/spacy/tests/pipeline/test_lemmatizer.py#L11-L25

The lemmatizer has default table lists for each mode:

https://github.com/explosion/spaCy/blob/8233359225df4b959a9d856ef208a66756f73502/spacy/pipeline/lemmatizer.py#L40-L53

If you write a custom lemmatizer, you can override it as needed (French `rule` uses `lookup` as a backoff):

https://github.com/explosion/spaCy/blob/8233359225df4b959a9d856ef208a66756f73502/spacy/lang/fr/lemmatizer.py#L17-L23

The Lemmatizer API is a bit clunky, but it was designed like this so you can add a new lemmatizer mode without implementing `__init__`, `initialize`, `__call__`, or serialization (well, unless you need data in a format other than tables), and it's intentionally all pure python. You should just need to add `[mode]_lemmatize` and `get_lookups_config` (only if necessary).

I'd recommend looking at the existing custom lemmatizers for more examples: `spacy/lang/*/lemmatizer.py`

**Reply by BramVanroy** [reply]
If I understand you correctly, the best way to go forward to modifying the lemmatizer would be to define a custom lemmatizer in Python (a separate python file). This file should then be referenced with `--code` during training and packaging so that `nlp.after_pipeline_creation` can replace the original lemmatizer with the custom lemmatizer? **Or** only changing the Lookups like so in a registered function.

```python
     nlp.config["initialize"]["components"]["lemmatizer"] = { 
         "lookups": {"@misc": "cope_lookups"} 
     } 
```

The first one seems slows as I think it first loads the default English() lemmatizer (and additional tables) and then would overwrite them with my custom ones. The second one might be too restrictive if I want to have a more expansive lemmatizer, like the French one that you posted. Is there a better way to use custom Classes for components in the config?

**Reply by adrianeboyd** [reply]
You can implement a custom language with its own lemmatizer factory, or you can use a custom lemmatizer component with its own factory name. In both cases you'd typically use `--code` to provide the code while training. Alternatively, you can put your custom registered functions in a python package and install that in your venv. The functions would be available through the entry points.

You would not use `after_creation` for either of these methods. Just define the lemmatizer as a pipeline component and provide the right `[components]` and `[initialize]` blocks. If you're only customizing the tables but not the algorithm, you just need the custom loader in the `[initialize]` block, which doesn't need to be packaged with the model later. Code for custom components or languages would need to be packaged with the model, or as a model dependency in a separate package.

**Reply by BramVanroy** [reply]
I think I got it. I will try some things out. and get back to you if I encounter related questions to lemmatisation. Thanks.
